File: crawlers/aggregator_hosts.py
"""Single source of truth for known aggregator/scam domains.

Loaded from data/aggregator_hosts.json (same file used by
lib/aggregatorHosts.ts). Imported by Discoverer + URL validators to
auto-reject rows whose apply_url host is on the blocklist.
"""

import json
import os
from urllib.parse import urlparse

_HOSTS_FILE = os.path.join(
    os.path.dirname(__file__), "..", "data", "aggregator_hosts.json"
)

with open(_HOSTS_FILE, encoding="utf-8") as f:
    _DATA = json.load(f)

AGGREGATOR_HOSTS: set[str] = set(_DATA["hosts"])


def is_aggregator_host(url: str) -> bool:
    """True if the given URL's host is on the aggregator/scam blocklist.
    Matches exact host AND subdomains (foo.mastersportal.com counts as
    mastersportal.com)."""
    if not url:
        return False
    try:
        host = (urlparse(url).hostname or "").lower().lstrip("www.")
        if host in AGGREGATOR_HOSTS:
            return True
        for blocked in AGGREGATOR_HOSTS:
            if host.endswith("." + blocked):
                return True
        return False
    except Exception:
        return False

File: lib/aggregatorHosts.ts
// Single source of truth for known aggregator/scam domains. Imported by:
//   - lib/supabase.ts (filters opportunity rendering on server queries)
//   - app/api/admin/feedback/auto-resolve/route.ts
//   - app/api/admin/sources/process/route.ts (Discoverer auto-process)
//
// Mirrors crawlers/aggregator_hosts.py. Both files read the same JSON so
// the list lives in exactly one place. Add new domains by editing
// data/aggregator_hosts.json; both languages pick it up automatically.

import aggregatorHostsData from "@/data/aggregator_hosts.json"

export const AGGREGATOR_HOSTS: Set<string> = new Set(
  (aggregatorHostsData as { hosts: string[] }).hosts
)

export function isAggregatorHost(url: string): boolean {
  if (!url) return false
  try {
    const host = new URL(url).hostname.toLowerCase().replace(/^www\./, "")
    if (AGGREGATOR_HOSTS.has(host)) return true
    for (const blocked of Array.from(AGGREGATOR_HOSTS)) {
      if (host.endsWith("." + blocked)) return true
    }
    return false
  } catch {
    return false
  }
}
